[MKC-667]benjamindannegard/nano33 TF tutorial update

BenjaminDannegard opened the pull request:
## What This PR Changes
- Updated Tensorflow tutorial since the library was removed from our library manager

## Contribution Guidelines
- [ ] I confirm that I have read the [contribution guidelines](https://github.com/arduino/docs-content/tree/main/contribution-templates) and comply with them.

BenjaminDannegard committed:
Changed instructions to be correct

BenjaminDannegard committed:
Updated image and instructions

BenjaminDannegard committed:
Updated more instructions

karlsoderby commented on content/hardware/03.nano/boards/nano-33-ble-sense/tutorials/get-started-with-machine-learning/get-started-with-machine-learning.md:
```diff
@@ -42,7 +42,7 @@ We’re excited to share some of the first examples and tutorials, and to see wh
 - Learn the fundamentals of TinyML implementation and training.
-- Use the [Arduino_LSM9DS1](https://www.arduino.cc/en/Reference/ArduinoLSM9DS1) and [Arduino_TensorFlowLite](https://www.arduino.cc/reference/en/libraries/arduino_tensorflowlite/) libraries
+- Use the [Arduino_LSM9DS1](https://www.arduino.cc/en/Reference/ArduinoLSM9DS1) and [TensorFlow Lite Micro Library for Arduino](https://github.com/tensorflow/tflite-micro-arduino-examples) libraries
 
```
```suggestion

***This tutorial was tested with [this version of the TensorFlow Lite Micro Library](https://github.com/tensorflow/tflite-micro-arduino-examples/archive/4056e5f13aafb4dd3eda46cec273133a1d5e9593.zip).***

```

karlsoderby committed:
Update content/hardware/03.nano/boards/nano-33-ble-sense/tutorials/get-started-with-machine-learning/get-started-with-machine-learning.md

Added a note stating that this tutorial works with a specific version of the TF library

per1234 commented on content/hardware/03.nano/boards/nano-33-ble-sense/tutorials/get-started-with-machine-learning/get-started-with-machine-learning.md:
```diff
@@ -131,10 +135,12 @@ We’ve adapted the tutorial below, so no additional hardware is needed – the
 ![Install Nano BLE board](assets/BoardsManager.png)
 
-**2.** Also, let's make sure we have all the libraries we need installed. If we are using the online IDE, there is no need to install anything. If we are using the offline IDE, this can be done by navigating to **Tools > Manage libraries...**, search for **Arduino_TensorFlowLite** and **Arduino:LSM9DS1**, and install them both. 
+**2.** Also, let's make sure we have all the libraries we need installed. If we are using the online IDE, there is no need to install anything. If we are using the offline IDE, this can be done by navigating to **Tools > Manage libraries...**, search for **Arduino:LSM9DS1**, and install it. 
```
```suggestion
**2.** Also, let's make sure we have the **Arduino_LSM9DS1** library installed. If we are using the online IDE, it is pre-installed. If we are using the offline IDE, this can be done by navigating to **Tools > Manage libraries...**, search for **Arduino_LSM9DS1**, and install it. 
```

The previous wording made it sound like step (3) could be skipped if using Arduino Web Editor.

I also propose a correction of a pre-existing error in the library name (it is "Arduino_LSM9DS1", not "Arduino:LSM9DS1", and you won't get any results from the search for the incorrect name).

per1234 commented on content/hardware/03.nano/boards/nano-33-ble-sense/tutorials/get-started-with-machine-learning/get-started-with-machine-learning.md:
```diff
@@ -131,10 +135,12 @@ We’ve adapted the tutorial below, so no additional hardware is needed – the
 
 ![Install the necessary libraries](assets/libManager.png)
 
+**3.** Lastly the [TensorFlow Lite Micro Library for Arduino](https://github.com/tensorflow/tflite-micro-arduino-examples) needs to be installed. You can find the installation instructions on the libraries github page [here](https://github.com/tensorflow/tflite-micro-arduino-examples#how-to-install).
```
```suggestion
**3.** Lastly the [TensorFlow Lite Micro Library for Arduino](https://github.com/tensorflow/tflite-micro-arduino-examples) needs to be installed. You can find the installation instructions on the library's GitHub page [here](https://github.com/tensorflow/tflite-micro-arduino-examples#how-to-install).
```

Minor grammar fix

per1234 reviewed (changes requested):
The IMU Classifier sketch at the end of the tutorial fails to compile when using the updated library version:

```text
C:\Users\per\Documents\Arduino\IMUClassifier\IMUClassifier.ino:23:10: fatal error: tensorflow/lite/version.h: No such file or directory
 #include <tensorflow/lite/version.h>
          ^~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.

exit status 1

Compilation error: tensorflow/lite/version.h: No such file or directory
```

As you can see, the library does not contain this file:

https://github.com/tensorflow/tflite-micro-arduino-examples/tree/4056e5f13aafb4dd3eda46cec273133a1d5e9593/src/tensorflow/lite

I also found there is a problem with the "**Setup Python Environment**" step of the Jupyter notebook:

```text
ERROR: Could not find a version that satisfies the requirement tensorflow==2.0.0-rc1 (from versions: 1.13.1, 1.13.2, 1.14.0, 1.15.0, 1.15.2, 1.15.3, 1.15.4, 1.15.5, 2.0.0, 2.0.1, 2.0.2, 2.0.3, 2.0.4, 2.1.0, 2.1.1, 2.1.2, 2.1.3, 2.1.4, 2.2.0, 2.2.1, 2.2.2, 2.2.3, 2.3.0, 2.3.1, 2.3.2, 2.3.3, 2.3.4, 2.4.0, 2.4.1, 2.4.2, 2.4.3, 2.4.4, 2.5.0, 2.5.1, 2.5.2, 2.5.3, 2.6.0rc0, 2.6.0rc1, 2.6.0rc2, 2.6.0, 2.6.0+zzzcolab20220506153740, 2.6.1, 2.6.2, 2.6.3, 2.6.4, 2.6.4+zzzcolab20220516125453, 2.6.5, 2.6.5+zzzcolab20220523104206, 2.7.0rc0, 2.7.0rc1, 2.7.0, 2.7.0+zzzcolab20220506150900, 2.7.1, 2.7.2, 2.7.2+zzzcolab20220516114640, 2.7.3, 2.7.3+zzzcolab20220523111007, 2.7.4, 2.8.0rc0, 2.8.0rc1, 2.8.0, 2.8.0+zzzcolab20220506162203, 2.8.1, 2.8.1+zzzcolab20220516111314, 2.8.1+zzzcolab20220518083849, 2.8.2, 2.8.2+zzzcolab20220523105045, 2.8.2+zzzcolab20220527125636, 2.8.2+zzzcolab20220629235552, 2.8.2+zzzcolab20220714152931, 2.8.2+zzzcolab20220714162028, 2.8.2+zzzcolab20220719082949, 2.8.2+zzzcolab20220822160911, 2.8.3, 2.9.0rc0, 2.9.0rc1, 2.9.0rc2, 2.9.0, 2.9.1, 2.9.2, 2.10.0rc0, 2.10.0rc1, 2.10.0rc2, 2.10.0rc3, 2.10.0)
ERROR: No matching distribution found for tensorflow==2.0.0-rc1
```

BenjaminDannegard committed:
Update content/hardware/03.nano/boards/nano-33-ble-sense/tutorials/get-started-with-machine-learning/get-started-with-machine-learning.md

Co-authored-by: per1234 <accounts@perglass.com>

BenjaminDannegard committed:
Update content/hardware/03.nano/boards/nano-33-ble-sense/tutorials/get-started-with-machine-learning/get-started-with-machine-learning.md

Co-authored-by: per1234 <accounts@perglass.com>

karlsoderby commented:
@BenjaminDannegard can you take a look at the comments from Per and test it out and adjust the content?

BenjaminDannegard commented:
> The IMU Classifier sketch at the end of the tutorial fails to compile when using the updated library version:
> 
> ```
> C:\Users\per\Documents\Arduino\IMUClassifier\IMUClassifier.ino:23:10: fatal error: tensorflow/lite/version.h: No such file or directory
>  #include <tensorflow/lite/version.h>
>           ^~~~~~~~~~~~~~~~~~~~~~~~~~~
> compilation terminated.
> 
> exit status 1
> 
> Compilation error: tensorflow/lite/version.h: No such file or directory
> ```
> 
> As you can see, the library does not contain this file:
> 
> https://github.com/tensorflow/tflite-micro-arduino-examples/tree/4056e5f13aafb4dd3eda46cec273133a1d5e9593/src/tensorflow/lite
> 
> I also found there is a problem with the "**Setup Python Environment**" step of the Jupyter notebook:
> 
> ```
> ERROR: Could not find a version that satisfies the requirement tensorflow==2.0.0-rc1 (from versions: 1.13.1, 1.13.2, 1.14.0, 1.15.0, 1.15.2, 1.15.3, 1.15.4, 1.15.5, 2.0.0, 2.0.1, 2.0.2, 2.0.3, 2.0.4, 2.1.0, 2.1.1, 2.1.2, 2.1.3, 2.1.4, 2.2.0, 2.2.1, 2.2.2, 2.2.3, 2.3.0, 2.3.1, 2.3.2, 2.3.3, 2.3.4, 2.4.0, 2.4.1, 2.4.2, 2.4.3, 2.4.4, 2.5.0, 2.5.1, 2.5.2, 2.5.3, 2.6.0rc0, 2.6.0rc1, 2.6.0rc2, 2.6.0, 2.6.0+zzzcolab20220506153740, 2.6.1, 2.6.2, 2.6.3, 2.6.4, 2.6.4+zzzcolab20220516125453, 2.6.5, 2.6.5+zzzcolab20220523104206, 2.7.0rc0, 2.7.0rc1, 2.7.0, 2.7.0+zzzcolab20220506150900, 2.7.1, 2.7.2, 2.7.2+zzzcolab20220516114640, 2.7.3, 2.7.3+zzzcolab20220523111007, 2.7.4, 2.8.0rc0, 2.8.0rc1, 2.8.0, 2.8.0+zzzcolab20220506162203, 2.8.1, 2.8.1+zzzcolab20220516111314, 2.8.1+zzzcolab20220518083849, 2.8.2, 2.8.2+zzzcolab20220523105045, 2.8.2+zzzcolab20220527125636, 2.8.2+zzzcolab20220629235552, 2.8.2+zzzcolab20220714152931, 2.8.2+zzzcolab20220714162028, 2.8.2+zzzcolab20220719082949, 2.8.2+zzzcolab20220822160911, 2.8.3, 2.9.0rc0, 2.9.0rc1, 2.9.0rc2, 2.9.0, 2.9.1, 2.9.2, 2.10.0rc0, 2.10.0rc1, 2.10.0rc2, 2.10.0rc3, 2.10.0)
> ERROR: No matching distribution found for tensorflow==2.0.0-rc1
> ```
@per1234 could we get in contact with the Tensorflow people and ask them if they can take a look at the library issue and fix it on their end? Do you know anyone over there we could talk to about this?

BenjaminDannegard committed:
Removed section that was not working

BenjaminDannegard committed:
Removed another section that caused error

BenjaminDannegard committed:
removed unused images

per1234 reviewed (changes requested):
This strategy of removing content instead of fixing it doesn't work.

The removed parts are essential components of the tutorial. With the removed parts, everything from the "Training a TensorFlow Lite Micro Model For Arduino" onward is nonsensical because it was all leading up to the culmination of the tutorial in the "Classifying IMU Data" section.

I'm confident that it will not be at all difficult to update the content. It is a very interesting and well written tutorial and I think performing the update would be a very enjoyable assignment.

RalphPuzon commented:
@BenjaminDannegard I manually added in version.h which I got from here: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/public/version.h

and my project compiled. if you're looking for some quick and dirty fix this may work for you.

per1234 reviewed (changes requested):
I'm adding a reminder here that there is now an additional copy of this content for the Nano 33 BLE Sense Rev2:

https://github.com/arduino/docs-content/blob/main/content/hardware/03.nano/boards/nano-33-ble-sense-rev2/tutorials/get-started-with-machine-learning/get-started-with-machine-learning.md

While the adjustments to the content are still in progress, it makes sense to work in only one of the two tutorials, but once the content adjustments are finalized in the Nano 33 BLE Sense version of the tutorial, they must also be applied to the Nano 33 BLE Sense Rev2 tutorial.

jhansson-ard commented:
@BenjaminDannegard @karlsoderby what is the status of this PR?

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/arduino/docs-content?pullRequest=556) <br/>All committers have signed the CLA.
